### ISSUE TITLE
Deeply nesting endpoints within multiple namespaces leads to large jumps in latency and cpu usage

### ISSUE THREAD

**Issue by atomless** [ruby-grape/grape#1416]
In my unit tests when I execute 10 simultaneous requests to a deeply nested endpoint of the form `':path_param1/namespaces1/:path_param2/namespace2/:path_param3/namespace3/namespace4/:path_param4'` it takes 60 seconds to complete and CPU consumed by Ruby peaks at over 90%. 

If I flatten that endpoint out and have it at the top level instead and have it named `post ':path_param1/namespaces1/:path_param2/namespace2/:path_param3/namespace3/namespace4/:path_param4' do` then 10 simultaneous requests to that endpoint complete in a few seconds and Ruby consumes only a few mb of CPU.

I will post some code to reproduce the issue here asap...

**Comment by atomless** [comment]
Hi @dblock I've tried to recreate a simple example with a single deeply nested endpoint, it is slower than when flat but not to the extent that I am seeing with our full API where there are lots of endpoints deeply nested at different levels in a single file. I have double checked and there does seem to be an issue with routing code in Grape as the deeply nested version of our api completes all requests without error but consumes 99% cpu and takes over a 80 seconds whereas the flat version takes under 10 seconds and consumes just 30% cpu. I'd be happy to send you a zip of our files if you think it would help but can't paste our proprietary code into this public space.

**Comment by dblock** [comment]
I think in order to get someone's attention who knows the router code well, like @namusyaka, you really need to reproduce this in a very small spec or project with a nice `benchmark do ...` around it.

**Comment by namusyaka** [comment]
Well I'm very interested in this issue. However, I couldn't reproduce the issue in my locally.
Probably, the issue can be avoided by **precompiling** all your routes.
I think we can provide a way for precompiling all routes. @dblock Thoughts?

**Comment by dblock** [comment]
Yes. But we shouldn't be guessing and benchmarking CPU vs. memory vs. time more systematically and not trying to optimize for someone's specific scenario.
